tapgarden/custodian.go

```diff
@@ -654,9 +663,91 @@ func (c *Custodian) checkProofAvailable(event *address.Event) error {
 // and pending address event. If a proof successfully matches the desired state
 // of the address, that completes the inbound transfer of an asset.
 func (c *Custodian) mapProofToEvent(p proof.Blob) error {
-	file := proof.NewEmptyFile(proof.V0)
-	if err := file.Decode(bytes.NewReader(p)); err != nil {
-		return fmt.Errorf("error decoding proof file: %w", err)
+	// We arrive here if we are notified about a new proof. The notification
+	// interface allows that proof to be a single transition proof. So if
+	// we don't have a full file yet, we need to fetch it now. The
+	// proof.NotifyArchiver.FetchProof() method will return the full file as
+	// per its Godoc.
+	var (
+		proofBlob = p
+		lastProof *proof.Proof
+		err       error
+	)
+	if !p.IsFile() {
+		log.Debugf("Received single proof, inspecting if matches event")
+		lastProof, err = p.AsSingleProof()
+		if err != nil {
+			return fmt.Errorf("error decoding proof: %w", err)
+		}
+
+		// Before we go ahead and fetch the full file, let's make sure
+		// we are actually interested in this proof.
+		matchesEvent := false
+		for _, event := range c.events {
```

Comment:
This is duplicated with the logic below.

Reply:
Yes. The first is to short-cut any calls that are made because of transfer proofs imported into the local universe that we might not be interested in anyway. Added some generic function magic to make it more compact.